File: python/0090_Subsets_II.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return [[]]
        
        nums.sort()
        subsets = []
        self.dfs(nums, 0, [False] * len(nums), [], subsets)
        return subsets
    
    def dfs(self, nums, index, visited, subset, subsets):
        subsets.append(list(subset))

        for i in range(index, len(nums)):
            if i > 0 and nums[i] == nums[i - 1] and visited[i - 1] == False:
                continue
            subset.append(nums[i])
            visited[i] = True
            self.dfs(nums, i + 1, visited, subset, subsets)
            visited[i] = False
            subset.pop()

File: cpp/0090_Subsets_II.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        vector<int> sub;
        dfs(nums, 0, sub, ans);
        return ans;
    }
    void dfs(vector<int>& nums, int i, vector<int>& sub, vector<vector<int>>& ans) {
        ans.push_back(sub);
        for (int j = i; j < nums.size(); j++) {
            if (j == i || nums[j] != nums[j - 1]) {
                sub.push_back(nums[j]);
                dfs(nums, j + 1, sub, ans);
                sub.pop_back();
            }
        }
    }
};